README.md
# Familiar Java Themes
This extension brings back the default color themes from your favorite Java IDEs,
such as **[Eclipse][1]** and **[IntelliJ IDEA][2]**! Currently a work in progress,
as syntax highlighting has only a been implemented for a few Java-related languages.

## Screenshots
*The screenshots below are taken with the [configuration tips][3] applied.*
<details>
  <summary>Eclipse Dark theme</summary>

  ![eclipse-dark](screenshots/eclipse-dark.png "Eclipse Dark theme")
</details>
<details>
  <summary>Eclipse Light theme</summary>

  ![eclipse-light](screenshots/eclipse-light.png "Eclipse Light theme")
</details>
<details>
  <summary>IntelliJ Dark theme</summary>

  ![intellij-dark](screenshots/intellij-dark.png "IntelliJ Dark theme")

  ![intellij-dark-terminal](screenshots/intellij-dark-terminal.png "ANSI terminal colors for the IntelliJ Dark theme")
</details>
<details>
  <summary>IntelliJ Light theme</summary>

  ![intellij-light](screenshots/intellij-light.png "IntelliJ Light theme")

  ![intellij-light-terminal](screenshots/intellij-light-terminal.png "ANSI terminal colors for the IntelliJ Light theme")
</details>
<details>
  <summary>IntelliJ Light (Classic) theme</summary>

  ![intellij-light-classic](screenshots/intellij-light-classic.png "IntelliJ Light (Classic) theme")

  ![intellij-light-classic-terminal](screenshots/intellij-light-classic-terminal.png "ANSI terminal colors for the IntelliJ Light (Classic) theme")
</details>

## Features
- Editor themes
  - Eclipse Dark
  - Eclipse Light
  - IntelliJ Dark (also known as "Darcula")
  - IntelliJ Light
  - IntelliJ Light (Classic)
  - IntelliJ+Eclipse Dark
    - Look and feel of IntelliJ with Eclipse Dark semantic coloring
- Matching themes for syntax highlighting
  - Support for **[semantic highlighting][4]**
  - Java syntax highlighting
  - XML syntax highlighting
  - JSON syntax highlighting
  - YAML syntax highlighting
  - Markdown syntax highlighting
  - EditorConfig syntax highlighting
  - More languages coming soon...
- Configuration helper to easily match the IntelliJ visuals

# Installing
### The extension can be installed from:
- [Visual Studio Marketplace][5]
- [Open VSX Registry][6]

### After installation
Select your preferred color theme from the VS Code settings, and optionally see
[configuration tips][3] for best results.

# Configuration tips
## General
It is **highly recommended** to use the semantic highlighting feature provided by version `0.69.0` or higher of the
[Language Support for Java(TM) by Red Hat][7] extension, in order to get the best results for Java syntax highlighting.
Previous versions of the Java extension do not provide all of the highlighting rules used by this theme extension.

Semantic highlighting for Java is controlled by the following VS Code settings:
- `editor.semanticHighlighting.enabled`
  - Defaults to true.
- `java.semanticHighlighting.enabled`
  - Defaults to true since version `0.66.0` of the Language Support for Java extension.
  - If you installed an earlier version of the Java extension, it's possible that this setting is still disabled.
  - You may need to restart VS Code after changing this setting in order for it to take effect.

Note that Semantic Highlighting needs to be computed by the Java Language Server,
which means that it comes with a slight delay and may not be immediately available when starting VS Code.

## IntelliJ themes
If you are used to IntelliJ IDEA, there are a couple of extra steps you can take
to make yourself feel at home. Note that these are just further configuration tips,
the color theme will work just fine without them.

#### Step 1: Install the JetBrains Mono font
The default font in IntelliJ IDEA is JetBrains Mono, which is quite different from
the default font in VS Code. If you want to use the JetBrains Mono font in VS Code,
you can follow [the instructions][8] for downloading and installing the font on your system.

#### Step 2: Configure font rendering and other visuals
Even if you were to use the JetBrains Mono font in VS Code, you might notice some
differences in how the font is rendered. The default settings in IntelliJ IDEA are
`16` for the font size and `1.2` for the line spacing. However, these values are not
calculated the same way in VS Code, which makes it tedious to match IDEA's font settings.
In addition, there are some visual features in IDEA that VS Code supports, but are
disabled by default and must be enabled manually in the settings.

In order to make your life easier, this extension provides the `Familiar Java Themes: Configuration Helper`
command, which can be found in the command palette (`Ctrl` + `Shift` + `P`). The command
automatically changes the correct VS Code settings to match the font and visual settings in IDEA.
It also gives you the ability to preview the changes, to see if you like them or not.
Note that even though the command tells VS Code to use the JetBrains Mono font, you still
have to install the font to your system yourself, via the intstructions in the step above.

If you prefer to tweak the settings yourself, you can find a detailed list of the settings
changed in the spoiler below.

<details>
  <summary>Settings changed by the configuration helper</summary>

  ```js
  {
    // Match IDEA's font settings for the editor
    "editor.fontFamily": "'JetBrains Mono'",
    "editor.fontSize": 16,
    "editor.lineHeight": 26,
    "editor.letterSpacing": 0.5,

    // Match IDEA's font settings for the terminal
    "terminal.integrated.fontFamily": "'JetBrains Mono'",
    "terminal.integrated.fontSize": 16,
    "terminal.integrated.lineHeight": 1.25,
    "terminal.integrated.letterSpacing": 1,

    // Match IDEA's font settings for the debug console
    "debug.console.fontFamily": "'JetBrains Mono'",
    "debug.console.fontSize": 16,
    "debug.console.lineHeight": 26,

    // The editor suggest widget (autocompletion)
    // has a smaller line height in IDEA
    "editor.suggestLineHeight": 23,

    // IDEA does not draw bold ANSI text in bright colors
    "terminal.integrated.drawBoldTextInBrightColors": false,

    // IDEA shows the current line in the editor gutter
    "editor.renderLineHighlight": "all"
  }
  ```
</details>

# Feedback
If you find missing colors, or ones that seem a bit off from what you are used to,
please [submit an issue][9] with screenshots showing what colors you expect and where.
Before you submit the issue, please check the known limitations below to see if your
suggestion is possible to implement.

<details>
  <summary>Known limitations</summary>

  - Semantic highlighting is currently only available in the editor of VS Code,
    and there is no way to enable it in the hover or suggest widget,
    which means that syntax colors will sometimes look strange in those widgets.
  - It is not possible to change the widths of borders and outlines in VS Code themes.
  - VS Code does not allow themes to modify all parts of the editor,
    which limits how well the themes can match their respective IDE.
</details>

If you like this extension, consider rating it on the [Marketplace][5]/[Registry][6].

# Acknowledgements
Credit goes to [Eclipse][1] and [IntelliJ IDEA][2] for the original color themes.

Versions used as reference:
- `Eclipse IDE for Java Developers` version `2020-06`
- `IntelliJ IDEA Community Edition` version `2020.2`

[1]: https://www.eclipse.org/eclipseide/
[2]: https://www.jetbrains.com/idea/
[3]: #configuration-tips
[4]: https://github.com/microsoft/vscode/wiki/Semantic-Highlighting-Overview
[5]: https://marketplace.visualstudio.com/items?itemName=zerodind.familiar-java-themes
[6]: https://open-vsx.org/extension/zerodind/familiar-java-themes
[7]: https://github.com/redhat-developer/vscode-java
[8]: https://www.jetbrains.com/lp/mono/#how-to-install
[9]: https://gitlab.com/zerodind/familiar-java-themes/-/issues


CHANGELOG.md
# Change Log

## [0.1.7] - 2022-09-04
- Support highlighting of Java Properties syntax by @d_scalzi

## [0.1.6] - 2022-08-16
- Add IntelliJ+Eclipse Dark theme by @d_scalzi

## [0.1.5] - 2021-10-23
- Add missing color for Java `super` keyword in generics
- Use new semantic tokens introduced in `redhat.java` version `1.0.0`
  - Fix colors for Java constructors

## [0.1.4] - 2021-08-01
- Fix wrong color for method parameters when not using semantic highlighting (#2)
- Add missing styling for bold/italic markdown in Eclipse themes

## [0.1.3] - 2021-02-20
- Fix broken colors for new semantic tokens added in `redhat.java` version `0.75.0`

## [0.1.2] - 2020-10-15
- Add missing color for italic markdown punctuation in IntelliJ themes
- Use new semantic tokens introduced in `redhat.java` version `0.69.0`
  - Fix colors for types in Java import declarations in Eclipse themes
  - Fix broken colors for Java modifier keywords
  - Add missing colors for Java generics in Eclipse Dark theme

## [0.1.1] - 2020-08-20
- Fix status bar colors in Eclipse themes

## [0.1.0] - 2020-08-18
- Initial release


extension.js
const vscode = require("vscode");

function activate(context) {
	context.subscriptions.push(vscode.commands.registerCommand(
		"familiarJavaThemes.configurationHelper",
		showConfigurationHelper
	));
}

const intelliJConfigurationValues = {
	// Match IDEA's font settings for the editor
	"editor.fontFamily": "'JetBrains Mono'",
	"editor.fontSize": 16,
	"editor.lineHeight": 26,
	"editor.letterSpacing": 0.5,

	// Match IDEA's font settings for the terminal
	"terminal.integrated.fontFamily": "'JetBrains Mono'",
	"terminal.integrated.fontSize": 16,
	"terminal.integrated.lineHeight": 1.25,
	"terminal.integrated.letterSpacing": 1,

	// Match IDEA's font settings for the debug console
	"debug.console.fontFamily": "'JetBrains Mono'",
	"debug.console.fontSize": 16,
	"debug.console.lineHeight": 26,

	// The editor suggest widget (autocompletion)
	// has a smaller line height in IDEA
	"editor.suggestLineHeight": 23,

	// IDEA does not draw bold ANSI text in bright colors
	"terminal.integrated.drawBoldTextInBrightColors": false,

	// IDEA shows the current line in the editor gutter
	"editor.renderLineHighlight": "all"
};

const CURRENT_VALUES = "Current Values";
const INTELLIJ_VALUES = "IntelliJ Values";
const DEFAULT_VALUES = "Default Values (Reset)";

function showConfigurationHelper() {
	const currentValues = {};
	Object.keys(intelliJConfigurationValues).forEach((key) => {
		currentValues[key] = vscode.workspace.getConfiguration(undefined, vscode.ConfigurationTarget.Global).get(key);
	});

	const onSelection = (item) => {
		switch (item) {
			case undefined:
			case CURRENT_VALUES: {
				setConfigurationValues(currentValues);
				break;
			}
			case INTELLIJ_VALUES: {
				setConfigurationValues(intelliJConfigurationValues);
				break;
			}
			case DEFAULT_VALUES: {
				setConfigurationValues({});
				break;
			}
		}
	}

	vscode.window.showQuickPick(
		[
			CURRENT_VALUES,
			INTELLIJ_VALUES,
			DEFAULT_VALUES
		],
		{
			placeHolder: "Select Configuration Values (Up/Down Keys to Preview)",
			ignoreFocusOut: true,
			onDidSelectItem: onSelection
		}
	).then(onSelection);
}

function setConfigurationValues(values) {
	Object.keys(intelliJConfigurationValues).forEach((key) => {
		vscode.workspace.getConfiguration().update(key, values[key], vscode.ConfigurationTarget.Global);
	})
}

module.exports = {
	activate
}
